File: src/privateboost/messages.py
"""Public dataclasses for protocol results."""

from dataclasses import dataclass
from typing import Literal

import numpy as np

from .crypto import Share

Loss = Literal["squared", "logistic"]


@dataclass
class CommittedStatsShare:
    """Stats share: y contains [x0, x0², x1, x1², ..., target, target²]."""

    commitment: bytes
    share: Share


@dataclass
class CommittedGradientShare:
    """Gradient share for a specific tree node."""

    round_id: int
    depth: int
    commitment: bytes
    share: Share
    node_id: int


@dataclass
class NodeTotals:
    """Gradient and hessian totals for a tree node."""

    gradient_sum: float
    hessian_sum: float


@dataclass
class GradientHistogram:
    """Gradient and hessian histograms for a feature."""

    gradient: np.ndarray
    hessian: np.ndarray


@dataclass
class BinConfiguration:
    """Histogram bin configuration for a feature."""

    feature_idx: int
    edges: np.ndarray
    inner_edges: np.ndarray
    n_bins: int


@dataclass
class MedianResult:
    """Result of median estimation for a feature."""

    feature_idx: int
    estimated_median: float
    histogram: np.ndarray
    bin_edges: np.ndarray


@dataclass
class SplitDecision:
    """A split decision for a tree node."""

    node_id: int
    feature_idx: int
    threshold: float
    gain: float
    left_child_id: int
    right_child_id: int
    g_left: float
    h_left: float
    g_right: float
    h_right: float


@dataclass
class LeafNode:
    """A leaf node with its prediction value."""

    node_id: int
    value: float
    n_samples: int

File: rust/privateboost/src/protocol/messages.rs
use crate::crypto::{Commitment, Share};

#[derive(Clone, Debug)]
pub enum Loss {
    Squared,
    Logistic,
}

pub struct CommittedStatsShare {
    pub commitment: Commitment,
    pub share: Share,
}

pub struct CommittedGradientShare {
    pub round_id: u64,
    pub depth: usize,
    pub commitment: Commitment,
    pub share: Share,
    pub node_id: usize,
}

#[derive(Clone, Debug)]
pub struct NodeTotals {
    pub gradient_sum: f64,
    pub hessian_sum: f64,
}

#[derive(Clone, Debug)]
pub(crate) struct GradientHistogram {
    pub gradient: Vec<f64>,
    pub hessian: Vec<f64>,
}

#[derive(Clone, Debug)]
pub struct BinConfiguration {
    pub feature_idx: usize,
    pub edges: Vec<f64>,
    pub inner_edges: Vec<f64>,
    pub n_bins: usize,
}

#[derive(Clone, Debug)]
pub struct SplitDecision {
    pub node_id: usize,
    pub feature_idx: usize,
    pub threshold: f64,
    pub gain: f64,
    pub left_child_id: usize,
    pub right_child_id: usize,
    pub g_left: f64,
    pub h_left: f64,
    pub g_right: f64,
    pub h_right: f64,
}
